From: Juho Vähä-Herttua <juhovh@iki.fi>
Date: Wed, 18 Jun 2008 12:41:29 +0000 (+0800)
Subject: OTHER: Add a hack to avcodec to avoid crashes coming from buggy ape decoder.
X-Git-Url: http://git.xmms.se/?p=xmms2-devel.git;a=commitdiff_plain;h=6f06f3409dfa82b7a3e7fdd682567d46fd65e262

OTHER: Add a hack to avcodec to avoid crashes coming from buggy ape decoder.
---

diff --git a/src/plugins/avcodec/avcodec.c b/src/plugins/avcodec/avcodec.c
index 28cb511..5f5760f 100644
--- a/src/plugins/avcodec/avcodec.c
+++ b/src/plugins/avcodec/avcodec.c
@@ -292,7 +292,7 @@ xmms_avcodec_read (xmms_xform_t *xform, xmms_sample_t *buf, gint len,
 		                                   &outbufsize, data->buffer,
 		                                   data->buffer_length);
 
-		if (bytes_read < 0) {
+		if (bytes_read < 0 || bytes_read > data->buffer_length) {
 			XMMS_DBG ("Error decoding data!");
 			return -1;
 		}
@@ -317,6 +317,8 @@ static gint64
 xmms_avcodec_seek (xmms_xform_t *xform, gint64 samples, xmms_xform_seek_mode_t whence, xmms_error_t *err)
 {
 	xmms_avcodec_data_t *data;
+	char outbuf[AVCODEC_MAX_AUDIO_FRAME_SIZE];
+	gint outbufsize, bytes_read = 0;
 	gint64 ret = -1;
 
 	g_return_val_if_fail (xform, -1);
@@ -325,6 +327,22 @@ xmms_avcodec_seek (xmms_xform_t *xform, gint64 samples, xmms_xform_seek_mode_t w
 	data = xmms_xform_private_data_get (xform);
 	g_return_val_if_fail (data, FALSE);
 
+	/* The buggy ape decoder doesn't flush buffers, so we need to finish decoding
+	 * the frame before seeking to avoid segfaults... this hack sucks */
+	while (data->buffer_length > 0) {
+		bytes_read = avcodec_decode_audio (data->codecctx, (short *) outbuf,
+		                                   &outbufsize, data->buffer,
+		                                   data->buffer_length);
+
+		if (bytes_read < 0 || bytes_read > data->buffer_length) {
+			XMMS_DBG ("Error decoding data!");
+			return -1;
+		}
+
+		data->buffer_length -= bytes_read;
+		g_memmove (data->buffer, data->buffer + bytes_read, data->buffer_length);
+	}
+
 	ret = xmms_xform_seek (xform, samples, whence, err);
 
 	if (ret >= 0) {
